Objects/unicodeobject.c

```diff
@@ -2163,12 +2163,27 @@ PyUnicode_FromStringAndSize(const char *u, Py_ssize_t size)
 PyObject *
 PyUnicode_FromString(const char *u)
 {
-    size_t size = strlen(u);
-    if (size > PY_SSIZE_T_MAX) {
-        PyErr_SetString(PyExc_OverflowError, "input too long");
-        return NULL;
+    /* Most calls of PyUnicode_FromString is for short ASCII string.
+     * So we optimize this function for them.
+     */
+    size_t len = 0;
+    int is_ascii = 1;
+
+    while (u[len] != '\0') {
+        if (u[len] > 127) {
+            is_ascii = 0;
```

Comment:
Can't you reuse fast _Py_bytes_isascii() somehow here? It works on unsigned long words (64 bits) rather than working on bytes (8 bits): it should be 8x faster.